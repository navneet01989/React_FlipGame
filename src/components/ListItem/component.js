import React from 'react';
import Animatedbasic from '../Animatedbasic';

import { setListData, updateCounter, recordPosition, refreshList, updatePairs } from '../../actions';
import { connect } from 'react-redux';
import { populateList, CARD_PAIRS_VALUE } from '../../utils';
import { Alert } from 'react-native';
  const ListItem = ({dispatch, data, counter, previousPositionClicked, updateListing, pairsCompleted, item, index}) => {
    const onCardPress = (item, index) => {
        const tempDataSource = data;
        if(tempDataSource[index].isShowing === false) {
            if(previousPositionClicked === -1) {
                tempDataSource[index].isShowing = true;
                dispatch(setListData(tempDataSource));
                dispatch(recordPosition(index));
            } else if(tempDataSource[previousPositionClicked].number === tempDataSource[index].number) {
                tempDataSource[index].isShowing = true;
                dispatch(setListData(tempDataSource));
                dispatch(recordPosition(-1));
                dispatch(updatePairs(pairsCompleted + 1));
                showAlertOnCompletion(pairsCompleted + 1);
            } else {
                tempDataSource[index].isShowing = true;
                setTimeout(() => {
                    tempDataSource[index].isShowing = false;
                    tempDataSource[previousPositionClicked].isShowing = false;
                    dispatch(recordPosition(-1));
                }, 800);
            }
            dispatch(updateCounter(counter + 1));
            dispatch(refreshList(!updateListing));
        }
    };
    
    const showAlertOnCompletion = (pairsCompleted) => {
        console.log('pairsCompleted', pairsCompleted, CARD_PAIRS_VALUE)
        if(pairsCompleted === CARD_PAIRS_VALUE) {
            Alert.alert(
                'Congratulation!',  
                `You won this game by ${counter + 1} steps`,  
                [  
                    {
                        text: 'Try another round', onPress: () => {
                        dispatch(setListData(populateList(CARD_PAIRS_VALUE)));
                        dispatch(updateCounter(0));
                        dispatch(updatePairs(0));
                    }
                },  
                ],
                {cancelable: false}   
            );
        }
    }

    return (
        <Animatedbasic testID="listItemParent" item={item} onPress={() => onCardPress(item, index)}/>
    )
  }
  const mapStateToProps = state => ({
    data: state.ListDataReducer.data,
    counter: state.ListDataReducer.counter,
    previousPositionClicked: state.ListDataReducer.previousPositionClicked,
    updateListing: state.ListDataReducer.updateListing,
    pairsCompleted: state.ListDataReducer.pairsCompleted,
  });
  export default connect(mapStateToProps)(ListItem);
