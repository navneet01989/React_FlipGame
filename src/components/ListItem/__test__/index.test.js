import React from 'react';
import renderer from 'react-test-renderer';
import ListItem from '../../../../src/components/ListItem';
import { Provider } from "react-redux";
import configureStore from 'redux-mock-store';
const mockStore = configureStore([]);

describe('Header snapshot', () => {
  let wrapper
  let store;
  beforeEach(() => {
    store = mockStore({
      ListDataReducer: {counter: 0}
    });
    const props = { item: {isShowing: true, number: 1}}
    wrapper = renderer.create(
      <Provider store={store}>
        <ListItem {...props}/>
      </Provider>
    );
  });
  afterEach(function() {
    store.clearActions()
  })
  test('render', () => {
    const result = wrapper.toJSON();
    expect(result).toMatchSnapshot();
    const card = wrapper.root.findByProps({ testID: 'listItem' });
    expect(card.type.displayName).toBe('Text');
  });
});