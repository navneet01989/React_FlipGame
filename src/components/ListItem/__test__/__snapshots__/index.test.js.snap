// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header snapshot render 1`] = `
<View
  accessible={true}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "blue",
      "flex": 1,
      "height": 250,
      "justifyContent": "center",
      "margin": 1,
      "opacity": 1,
      "transform": Array [
        Object {
          "rotate": "360deg",
        },
      ],
    }
  }
>
  <Text
    style={
      Object {
        "color": "white",
        "fontSize": 25,
      }
    }
    testID="listItem"
  >
    1
  </Text>
</View>
`;
